_posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd

```diff
@@ -0,0 +1,492 @@
+
+```{r}
+r2_act$representations[[2]](elem_1)
+```
+
+```         
+             [,1]          [,2]
+[1,] 6.123234e-17 -1.000000e+00
+[2,] 1.000000e+00  6.123234e-17
+```
+
+This is the so-called standard representation,
+
+$$
+\begin{bmatrix} \cos(\theta) & -\sin(\theta) \\ \sin(\theta) & \cos(\theta) \end{bmatrix}
+$$
+
+, evaluated at $\theta = \pi/2$. (It is called the standard representation because it directly comes from how the group is defined (namely, a rotation by $\theta$ in the plane).
+
+The other interesting representation to point out is the fourth: the only one that's not irreducible.
+
+```{r}
+r2_act$representations[[4]](elem_1)
+```
+
+```         
+[1,]  5.551115e-17 -5.551115e-17 -8.326673e-17  1.000000e+00
+[2,]  1.000000e+00  5.551115e-17 -5.551115e-17 -8.326673e-17
+[3,]  5.551115e-17  1.000000e+00  5.551115e-17 -5.551115e-17
+[4,] -5.551115e-17  5.551115e-17  1.000000e+00  5.551115e-17
+```
+
+This is the so-called *regular* representation. The regular representation acts via permutation of group elements, or, to be more precise, of the basis vectors that make up the matrix. Obviously, this is only possible for finite groups like $C_n$, since otherwise there'd be an infinite amount of basis vectors to permute.
+
+To better see the action encoded in the above matrix, we clean up a bit:
+
+```{r}
+round(r2_act$representations[[4]](elem_1))
+```
+
+```         
+    [,1] [,2] [,3] [,4]
+[1,]    0    0    0    1
+[2,]    1    0    0    0
+[3,]    0    1    0    0
+[4,]    0    0    1    0
+```
+
+This is a step-one shift to the right of the identity matrix. The identity matrix, mapped to element 0, is the non-action; this matrix instead maps the zeroth action to the first, the first to the second, the second to the third, and the third to the first.
+
+We'll see the regular representation used in a neural network soon. Internally -- but that need not concern the user -- *escnn* works with its decomposition into irreducible matrices. Here, that's just the bunch of irreducible representations we saw above, numbered from one to three.
+
+Having looked at how groups and representations figure in `escnn`, it is time we approach the task of building a network.
+
+## Representations, for real: `escnn$nn$FieldType`
+
+So far, we've characterized the input space ($\mathbb{R}^2$), and specified the group action. But once we enter the network, we're not in the plane anymore, but in a space that has been extended by the group action. Rephrasing, the group action produces *feature vector fields* that assign a feature vector to each spatial position in the image.
+
+Now we have those feature vectors, we need to specify how they transform under the group action. This is encoded in an `escnn$nn$FieldType` . Informally, we could say that a field type is the *data type* of a feature space. In defining it, we indicate two things: the base space, a `gspace`, and the representation type(s) to be used.
```

Comment:
> Now we have those feature vectors, we need to specify how they transform under the group action.

Again, this is really helpful and clear! _This_ is 100% the type of writing I'm referring to when I talk about helping a reader to avoid getting lost! 🙂

Reply:
I'm very happy to hear that, thank you!